src/snowflake/cli/api/metrics.py

```diff
@@ -54,17 +67,125 @@ class CLICounterField:
+    PARENT_KEY = "parent"
+    PARENT_ID_KEY = "parent_id"
+    START_TIME_KEY = "start_time"
+    EXECUTION_TIME_KEY = "execution_time"
+    ERROR_KEY = "error"
+
+    def __init__(self, name: str, parent: _CLIMetricsStep | None = None):
+        if not name:
+            raise CLIMetricsInvalidUsageError("step name must not be empty")
+
+        # to disambiguate steps with same name
+        self._step_id: int = next(self._id_counter)
+        self._name: str = name
+
+        self._parent: Optional[str] = parent.name if parent is not None else None
+        self._parent_id: Optional[int] = parent.step_id if parent is not None else None
+
+        self._monotonic_start: Optional[
+            float
+        ] = None  # monotonic time for execution time
+        self._start_time: Optional[float] = None  # timestamp
+        self._execution_time: Optional[float] = None
+
+        self._error: Optional[str] = None
+
+    def start(self) -> None:
+        if self._monotonic_start:
+            raise CLIMetricsInvalidUsageError("step has already started")
+        self._start_time = time.time()
+        self._monotonic_start = time.monotonic()
+
+    def end(self, error: Optional[BaseException] = None) -> None:
+        if self._execution_time:
+            raise CLIMetricsInvalidUsageError("step has already ended")
+        if not self._monotonic_start:
+            raise CLIMetricsInvalidUsageError("step has not started")
+
+        if error:
+            self._error = type(error).__name__
+
+        self._execution_time = time.monotonic() - self._monotonic_start
+
+    @property
+    def step_id(self) -> int:
+        return self._step_id
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def parent(self) -> Optional[str]:
+        return self._parent
+
+    @property
+    def parent_id(self) -> Optional[int]:
+        return self._parent_id
+
+    @property
+    def start_time(self) -> float:
+        if not self._start_time:
+            raise CLIMetricsInvalidUsageError("start_time accessed before step started")
+        return self._start_time
+
+    @property
+    def execution_time(self) -> float:
+        if not self._execution_time:
+            raise CLIMetricsInvalidUsageError(
+                "execution_time accessed before step ended"
+            )
+        return self._execution_time
+
+    @property
+    def error(self) -> Optional[str]:
+        """
+        name of the error class that was raised during the execution of this step
+        """
+        return self._error
+
+    def to_dict(self) -> Dict[str, Union[str, int, float, None]]:
+        return {
+            self.ID_KEY: self.step_id,
+            self.NAME_KEY: self.name,
+            self.PARENT_KEY: self.parent,
+            self.PARENT_ID_KEY: self.parent_id,
+            self.START_TIME_KEY: self.start_time,
+            self.EXECUTION_TIME_KEY: self.execution_time,
+            self.ERROR_KEY: self.error,
+        }
```

Comment:
can this class be an `@dataclass`? then we wouldn't need this method or all the getters

Reply:
Converted to dataclass

Regarding your suggestion to implement `to_dict` as `dict((field.name, getattr(obj, field.name)) for field in fields(obj) if field.name in FIELDS_TO_SERIALIZE)`, I think that would add additional complexity and make it harder to reason about, since we also add additional keys to the dict that are extracted (i.e. parent name, parent id, error name) from the dataclass fields. I think it's simpler to be explicit here since there aren't too many keys

Let me know if you have other thoughts!